### PR TITLE
Module overview 3

### PR DESCRIPTION
**Pull request by abea**
Yet more module overview sections

- helpers
- extendHelpers
- restApiRoutes
- extendRestApiRoutes
- apiRoutes
- extendApiRoutes
- renderRoutes
- routes
- handlers
- extendHandlers

We might want to do some reordering later, but I didn't want to make it more confusing to review.

### CODE REVIEW

File: docs/reference/module-api/module-overview.md
Comment (boutell) [inline docs/reference/module-api/module-overview.md:746]: How about: "the functions must generate a response via `res` to avoid leaking resources, typically using the `res.redirect` or `res.send` methods." There are other, less often used ways to end a response, such as calling `res.send`, so I don't want to make it sound like Apostrophe is doing anything other than letting you do as you like with Express here.

File: docs/reference/module-api/module-overview.md
Comment (boutell) [inline docs/reference/module-api/module-overview.md:774]: Event handlers may be async functions, and they will be awaited before execution continues.

**Review by boutell — CHANGES_REQUESTED** [review]
Small stuffs. This is so 💯

File: docs/reference/module-api/module-overview.md
Comment (localghost443) [inline docs/reference/module-api/module-overview.md:487]: "The `helpers` section" could probably be dropped. We're already there.

File: docs/reference/module-api/module-overview.md
Comment (localghost443) [inline docs/reference/module-api/module-overview.md:543]: "**Those route handlers can be totally overridden** by adding identically named functions **to object** returned by the `restApiRoutes` function. If you **simply wish to add to** the existing behavior of the REST API routes, see [`extendRestApiRoutes`](#extendrestapiroutes-self)."

"Those route handlers can be totally overridden" -> "You can completely override these route handlers" (completely optional?)

"to object" -> "to objects"

"If you **simply wish to add to**" -> "To add to and extend existing behavior"

File: docs/reference/module-api/module-overview.md
Comment (boutell) [reply docs/reference/module-api/module-overview.md:746]: `routes` should be used only where `apiRoutes` or `restApiRoutes` or `renderRoutes` won't do, such as when you need to use `res.redirect` or pipe a stream. We should state this clearly to help guide people to the painless solution. Failing to do the right thing with `res` or handle errors etc. is a common cause of Node.js bugs, which is why apiRoutes exists.

(At some point I'd like to add a way to use `res.redirect` via a return value. Piping a stream will probably always be an "uh... just use routes()" scenario.")

File: docs/reference/module-api/module-overview.md
Comment (localghost443) [inline docs/reference/module-api/module-overview.md:576]: "in the `extendRestApiRoutes` section" -> remove

File: docs/reference/module-api/module-overview.md
Comment (boutell) [inline docs/reference/module-api/module-overview.md:543]: You imply it, but this really needs to be emphasized:

All piece type modules, as well as the `@apostrophecms/page` module, already have a complete REST API and you should not use `restApiRoutes` with these. The most likely use of `restApiRoutes` in a project would be to provide a custom REST API to a database or service not already built into Apostrophe. `restApiRoutes` is also not for custom routes that don't map to one of the standard REST URLs. If you need to add a custom route over and above the REST API of pieces or pages, you should do that with `apiRoutes`.

File: docs/reference/module-api/module-overview.md
Comment (localghost443) [inline docs/reference/module-api/module-overview.md:636]: `cheapest` could probably be replaced by something that translates a bit better, since its pretty much slang. Using product....my first thought is `brand/s` or a category like `denim-jacket/s`. Denim and jacket are both fairly universal, while being very American!

Bonus points if we find a way to use infant bodysuit

File: docs/reference/module-api/module-overview.md
Comment (localghost443) [inline docs/reference/module-api/module-overview.md:683]: "in the `extendApiRoutes` section" -> dont need (imo)

File: docs/reference/module-api/module-overview.md
Comment (localghost443) [inline docs/reference/module-api/module-overview.md:711]: "the module as an argument" -> is this the module name?

File: docs/reference/module-api/module-overview.md
Comment (localghost443) [inline docs/reference/module-api/module-overview.md:774]: "simply by name" -> "by name"

File: docs/reference/module-api/module-overview.md
Comment (localghost443) [reply docs/reference/module-api/module-overview.md:543]: I love wishes tho

File: docs/reference/module-api/module-overview.md
Comment (abea) [reply docs/reference/module-api/module-overview.md:679]: I'm holding off on full documentation of that here since it could be a big chunk. I'm going to link over to it once available instead. There'll be a sentence like "See more information on using the `error()` method."

File: docs/reference/module-api/module-overview.md
Comment (abea) [reply docs/reference/module-api/module-overview.md:746]: I've added a new :::tip block a few lines down.

File: docs/reference/module-api/module-overview.md
Comment (abea) [reply docs/reference/module-api/module-overview.md:774]: I've been internally debating how much to explicitly say that things can be async. It's probably a good idea to do more.

File: docs/reference/module-api/module-overview.md
Comment (abea) [reply docs/reference/module-api/module-overview.md:543]: > The most likely use of restApiRoutes in a project of your own would be to provide a custom REST API to a database or service not already built into Apostrophe.

@boutell Could you suggest a more specific use case? I want to make sure I'm communicating what you're saying correctly. Like I create a `rando` module that is not a piece, then I add a REST API that actually hits a third-party database?

Comment (abea) [reply docs/reference/module-api/module-overview.md:543]: Maybe `apiRoutes` should appear first in the page.

Comment (abea) [reply docs/reference/module-api/module-overview.md]: I'm redoing that whole section based on Tom's notes.

File: docs/reference/module-api/module-overview.md
Comment (abea) [reply docs/reference/module-api/module-overview.md:584]: These should say "function" not "section."

Comment (abea) [reply docs/reference/module-api/module-overview.md:584]: Or nothing.

File: docs/reference/module-api/module-overview.md
Comment (abea) [reply docs/reference/module-api/module-overview.md:644]: 😕 "cheapest" simply means "most cheap." Least expensive.

File: docs/reference/module-api/module-overview.md
Comment (abea) [reply docs/reference/module-api/module-overview.md:719]: No, it's the actual module. That's what `self` is.

File: docs/reference/module-api/module-overview.md
Comment (localghost443) [inline docs/reference/module-api/module-overview.md:759]:
```suggestion
We recommend using `apiRoutes` or `restApiRoutes` whenever possible before using `routes`, as they handle potential pitfalls of Express routes. There are situations where writing Express routes may be necessary, such as when you need to use `res.redirect` or pipe a stream.
```

File: docs/reference/module-api/module-overview.md
Comment (localghost443) [inline docs/reference/module-api/module-overview.md:786]:
```suggestion
Events belonging to the same module where the handlers are defined, or from its base class, can be referenced by name, e.g., `beforeInsert` for any piece type. You may also add `handlers` in one module that respond to events in other modules. Those event names should be prefixed with the name of the module where the event occurs, followed by a colon, e.g., `@apostrophecms/page:beforeSend`.
```

File: docs/reference/module-api/module-overview.md
Comment (boutell) [reply docs/reference/module-api/module-overview.md:487]: How about: "the individual helper methods may accept arguments when called from your templates."

File: docs/reference/module-api/module-overview.md
Comment (boutell) [reply docs/reference/module-api/module-overview.md:693]: "Your extended API route function should"?

**Review by boutell — APPROVED** [review]
I'm approving with suggestions on Ben's grammar notes.

File: docs/reference/module-api/module-overview.md
Comment (abea) [reply docs/reference/module-api/module-overview.md:786]: "handlers" is also the name of the type of function, like we call then "Event handlers" earlier.

**Review by localghost443 — APPROVED** [review]

**Pull request merged**
